File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        temp = []
        self.combinations(0, ans, temp, target, candidates)
        return ans

    def combinations(self, ind, arr, temp, target, candidates):

        if target == 0:
            arr.append(temp[:])
            return

        if target < 0 or ind == len(candidates):
            return

        target -= candidates[ind]
        temp.append(candidates[ind])
        self.combinations(ind, arr, temp, target, candidates)

        target += candidates[ind]
        temp.pop()
        self.combinations(ind+1, arr, temp, target, candidates)

File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const ds = [];
    const ans = [];
    function recurs(ind, candidates, target, ds, ans){
        if(ind == candidates.length){
            if (target == 0){
                ans.push(ds.slice(0,ds.length));
            }
            return;
        }
        if(candidates[ind] <= target){
            ds.push(candidates[ind]);
            recurs(ind, candidates, target-candidates[ind], ds, ans);
            ds.pop();
        }
        recurs(ind+1, candidates, target, ds, ans);
    }
    recurs(0, candidates, target, ds, ans);
    return ans;
};